core-api/batch/prime-finder/src/main/java/PrimeFinder.java

```diff
@@ -121,38 +114,41 @@ private static boolean isPrime(int n) {
         @Override
         public void init(@Nonnull Context context) {
             totalParallelism = context.totalParallelism();
-            localParallelism = context.localParallelism();
         }
 
 
         @Override @Nonnull
         public Function<Address, ProcessorSupplier> get(@Nonnull List<Address> addresses) {
-            Map<Address, ProcessorSupplier> map = new HashMap<>();
-            for (int i = 0; i < addresses.size(); i++) {
-                Address address = addresses.get(i);
-                int start = i * localParallelism;
-                int end = (i + 1) * localParallelism;
-                int mod = totalParallelism;
-                map.put(address, count -> range(start, end)
-                        .mapToObj(index -> new NumberGenerator(range(0, limit).filter(f -> f % mod == index)))
-                        .collect(Collectors.toList())
-                );
-            }
-            return map::get;
+            return address -> count -> IntStream.range(0, count)
+                                                .mapToObj(i -> new NumberGenerator(limit, totalParallelism))
+                                                .collect(Collectors.toList());
         }
     }
 
     static class NumberGenerator extends AbstractProcessor {
 
-        private final Traverser<Integer> traverser;
+        private final int limit;
+        private final int totalParallelism;
+        private int nextValue;
+
+        NumberGenerator(int limit, int totalParallelism) {
+            this.limit = limit;
+            this.totalParallelism = totalParallelism;
+        }
 
-        NumberGenerator(IntStream stream) {
-            traverser = traverseStream(stream.boxed());
+        @Override
+        protected void init(@Nonnull Context context) {
+            nextValue = context.globalProcessorIndex();
```

Comment:
wouldn't it be cleaner maybe just to create a traverser here? that returns nextValue += totalParallelism on `next` ?